api: gpio: add gpio direction query functions and driver api

cfriedt opened the pull request:
Some services require the ability to query whether a GPIO pin is configured as an input, output, both, or neither prior to performing any operations at the service level. This is done in order to reduce state tracking within the service.

To that end, this change adds
* `gpio_port_get_direction()`
* `gpio_pin_is_input()`, and
* `gpio_pin_is_output()`

Fixes #26938

* Added tests under `tests/drivers/gpio_get_direction` and implementations for a handful of drivers.
* Add myself and @henrikbrixandersen as collaborators for GPIO
* Remove `(void)` casts and use `__unused` for variables unused due to disabled assertions.

cfriedt commented:
Implementations forthcoming for `gpio_emul` and `gpio_cc13xx_cc26xx`..

mmahadevan108 commented on drivers/gpio/Kconfig:
```diff
@@ -15,6 +15,11 @@ module = GPIO
 source "subsys/logging/Kconfig.template.log_config"
 
+config GPIO_GET_DIRECTION_BITS
+	bool
```
Why do we need this Kconfig?

cfriedt replied:
To indicate whether a driver supports getting pin direction. Support for this feature was requested to be optional in a previous iteration of this change from a long while ago

mmahadevan108 replied:
If the API returns `ENOTSUP`, wouldn't that indicate whether this is supported or not supported?

cfriedt replied:
I would prefer to remove it as well, personally, but I think with the GPIO API, it would require a fair amount of work to implement the feature for each platform. In a lot of applications, this capability isn't necessary either, so making it configurable would reduce some amount of code footprint.

cfriedt replied:
@mmahadevan108 - you do bring up a good point though. One thing I had not really considered was the case where are are multiple GPIO drivers - e.g. the SoC GPIO driver and then something like a GPIO expander. Globally, the API might be enabled, but still not all platforms would be able to support it.

cfriedt commented:
@mmahadevan108 - reworked. It's a bit cleaner now. Used `IS_ENABLED()` rather than `#ifdef`s everywhere. `CONFIG_GPIO_GET_DIRECTION` is now user-configurable, and there is a runtime check within the syscall for support (so that multiple drivers on the same system can co-exist with support / lack of support - e.g. in the case of local GPIO and a GPIO expander).

cfriedt commented:
* fix typos in `gpio_handlers.c`

cfriedt commented:
@mnkp - can you please review? @mmahadevan108 - I think I've addressed your concerns, so please take another look.

mbolivar-nordic reviewed (dismissed):
The approach and API changes look good to me. I didn't check the drivers.

mmahadevan108 commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +754,96 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+						      gpio_port_pins_t *inputs,
+						      gpio_port_pins_t *outputs)
+{
+	if (!IS_ENABLED(CONFIG_GPIO_GET_DIRECTION)) {
+		return -ENOTSUP;
+	}
```
Do we need this as lines 784-786 has a similar check.

cfriedt replied:
Yeah, could probably remove it. I guess technically if each driver uses `IS_ENABLED()` in the driver API, then it's not needed.

mmahadevan108 replied:
Apologies to bring this up again, but why do we need the Kconfig `GPIO_GET_DIRECTION`
Shouldn't the below check be able to address platforms that do not implement the new API to get the GPIO direction.
```
       if (api->port_get_direction_raw == NULL) {
               return -ENOTSUP;
       }
```

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -563,6 +563,8 @@ __subsystem struct gpio_driver_api {
 			       struct gpio_callback *cb,
 			       bool set);
 	uint32_t (*get_pending_int)(const struct device *dev);
+	int (*port_get_direction_raw)(const struct device *port, gpio_port_pins_t *inputs,
+					   gpio_port_pins_t *outputs);
```
To avoid impacting the footprint when not enabled, the API struct function pointer (along with the implementation functions in all drivers) should be make conditional as shown below. This is how conditional APIs are handled in other Zephyr driver classes.

```suggestion
#ifdef CONFIG_GPIO_GET_DIRECTION
	int (*port_get_direction_raw)(const struct device *port, gpio_port_pins_t *inputs,
					   gpio_port_pins_t *outputs);
#endif /* CONFIG_GPIO_GET_DIRECTION */
```

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +754,92 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ * @retval -EWOULDBLOCK if operation would block.
+ */
+__syscall int gpio_port_get_direction(const struct device *port, gpio_port_pins_t *inputs,
+					   gpio_port_pins_t *outputs);
+
+static inline int z_impl_gpio_port_get_direction(const struct device *port,
+						      gpio_port_pins_t *inputs,
+						      gpio_port_pins_t *outputs)
```
The syscalls must always be defined, but the `z_impl_*` and `z_vrfy_*` functions shall be condtional on `CONFIG_GPIO_GET_DIRECTION` (see https://docs.zephyrproject.org/latest/kernel/usermode/syscalls.html#implementation-details).

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +754,92 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+	const struct gpio_driver_api *api = (const struct gpio_driver_api *)port->api;
+
+	if (api->port_get_direction_raw == NULL) {
+		return -ENOTSUP;
```
I think the correct return value here would be `-ENOSYS` (see https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes).

henrikbrixandersen commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -81,11 +81,39 @@ static int setup(void)
 	zassert_equal(rc, 0,
 		      "pin config input failed");
 
+	if (IS_ENABLED(CONFIG_GPIO_GET_DIRECTION)) {
```
Perhaps move these into their own test case instead of piggybacking on setup - and properly indicate if they are not tested by skipping the test?

henrikbrixandersen replied:
I see you have moved this to a completely new test suite? That's perhaps overkill. I merely suggested a new, dedicated test case (function) in the `gpio_basic_api` test suite. Either way works, though.

henrikbrixandersen commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -261,6 +261,25 @@ static void gpio_mcux_port_isr(const struct device *dev)
+		*inputs = ~gpio_base->PDDR;
+	}
+
+	if (inputs != NULL) {
+		*outputs = gpio_base->PDDR;
+	}
```
This doesn't look right? inputs vs. outputs?

cfriedt replied:
Fixed!

cfriedt commented:
@henrikbrixandersen - yeah, i feel that if the function isn't supported in all drivers, it can't technically be a part of the basic api. Can always move in at a later date if it becomes more widely adopted.

cfriedt commented:
* added `[EXPERIMENTAL]` and `select EXPERIMENTAL` to Kconfig

henrikbrixandersen commented:
Seems we have two very related PRs for extending the GPIO API with query functions. This one and #43225.

cfriedt commented:
* assert `port` is not `NULL` in tests
* remove `_raw` from `port_get_direction`

trentp-igor commented:
The way  `gpio_pin_is_output/input()` loop over the entire port only to throw away all pins but one seems really inefficient.  Since is goes through a function pointer the compiler won't be able to optimize away any of it.

Why not have `port_get_direction()` take a mask of which pins to query?  This makes `gpio_pin_is_in/output()` much more efficient.
```
/// @param pins Set of pins to query the direction of.  The value of input and output for bits corresponding to those NOT set in pins is undefined.
__syscall int gpio_port_get_pins_direction(const struct device *port, gpio_port_pins_t pins,
                                           gpio_port_pins_t *inputs, gpio_port_pins_t *outputs);

static inline int gpio_pin_is_input(const struct device *port, gpio_pin_t pin)
{
	…
	rv = gpio_port_get_pins_direction(port, BIT(pin), &pins, NULL);

	return !!((gpio_port_pins_t)BIT(pin) & pins);
}
```
If you want to get all the pins, `gpio_port_get_direction()` is a trivial wrapper:
```
static inline int gpio_port_get_direction(const struct device *port, gpio_port_pins_t *inputs, gpio_port_pins_t *outputs)
{
        return gpio_port_get_pins_direction(port, (gpio_port_pins_t)-1, inputs, output);
}
```

The implementations that already have the direction in a bitfield don't need to change.  Those that need loops only get slightly more complex.  Example from gpio_emul.c:
```
pins &= BIT_MASK(config->num_pins);
for (int i = find_lsb_set(pins) - 1; pins; pins &= pins - 1, i = find_lsb_set(pins) - 1) {
        ip |= !!(drv_data->flags[i] & GPIO_INPUT) * BIT(i);
}
*input = ip;
```
Example of compiled code, https://godbolt.org/z/foq7rcsMx

cfriedt commented:
> The way `gpio_pin_is_output/input()` loop over the entire port only to throw away all pins but one seems really inefficient. Since is goes through a function pointer the compiler won't be able to optimize away any of it.

True - I thought about this before as well. Eventually, I settled on simplicity because platforms vary in the way they access this information (e.g. some only require 1 read-register instruction, some HALs require looping, and ffs can be expensive on some platforms).

> Why not have `port_get_direction()` take a mask of which pins to query? This makes `gpio_pin_is_in/output()` much more efficient.

I think that's fine at the API level, and yes, the implementation is fairly trivial.

> The implementations that already have the direction in a bitfield don't need to change. Those that need loops only get slightly more complex. Example from gpio_emul.c:
> 
> ```
> pins &= BIT_MASK(config->num_pins);
> for (int i = find_lsb_set(pins) - 1; pins; pins &= pins - 1, i = find_lsb_set(pins) - 1) {
>         ip |= !!(drv_data->flags[i] & GPIO_INPUT) * BIT(i);
> }

Sure, that works.

cfriedt commented:
* fixed typo

Commit:
api: gpio: add gpio direction query functions and driver api

Some services require the ability to query whether a GPIO pin
is configured as an input, output, both, or neither prior to
performing any operations at the service level. This is done
in order to reduce state tracking within the service.

To that end, this change adds
* `gpio_port_get_direction()`
* `gpio_pin_is_input()`, and
* `gpio_pin_is_output()`

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
tests: drivers: gpio: tests for gpio_port_get_direction

Tests for gpio_port_get_direction(), gpio_pin_is_input(),
and gpio_pin_is_output().

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
drivers: gpio_emul: support gpio_port_get_direction

Support querying the direction of emulated GPIO.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
drivers: gpio: cc13xx_cc26xx: support port_get_direction

Support querying the direction of GPIO.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
drivers: gpio: eos_s3: support port_get_direction

Support querying the direction of GPIO.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
drivers: gpio: litex: support port_get_direction

Support querying the direction of GPIO.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
drivers: gpio: sifive: support port_get_direction

Support querying the direction of GPIO.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
drivers: gpio: mcux: support port_get_direction

Support querying the direction of GPIO.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

Commit:
MAINTAINERS: gpio: add collaborators

Adding myself and Henrik Brix Andersen as collaborators on GPIO.

Signed-off-by: Christopher Friedt <cfriedt@fb.com>

cfriedt commented:
@mnkp, @henrikbrixandersen - does this look acceptable now? Was discussed in last week's dev-review, and the consensus was that there should be room for both this and #43225

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +756,97 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+{
+	int rv;
+	gpio_port_pins_t pins;
+	const struct gpio_driver_config *cfg = (const struct gpio_driver_config *)port->config;
+
+	(void)cfg;
```
```suggestion
	__unused const struct gpio_driver_config *cfg = (const struct gpio_driver_config *)port->config;

```

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +756,97 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+ * @param inputs Pointer to a variable where input directions will be stored.
+ * @param outputs Pointer to a variable where output directions will be stored.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing an external GPIO chip#.
+ * @retval -EWOULDBLOCK if operation would block.
```
```suggestion
 * @retval 0 If successful.
 * @retval -EIO I/O error when accessing an external GPIO chip.
 * @retval -EWOULDBLOCK if operation would block.
```

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +756,97 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+{
+	int rv;
+	gpio_port_pins_t pins;
+	const struct gpio_driver_config *cfg = (const struct gpio_driver_config *)port->config;
+
+	(void)cfg;
```
Same as above.

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +756,97 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+		return rv;
+	}
+
+	return !!((gpio_port_pins_t)BIT(pin) & pins);
```
Same as above.

cfriedt replied:
I can add a cast if you like.

henrikbrixandersen reviewed (changes requested):
Looks good, few minor observations:

cfriedt commented on include/zephyr/drivers/gpio.h:
```diff
@@ -752,6 +756,97 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
+		return rv;
+	}
+
+	return !!((gpio_port_pins_t)BIT(pin) & pins);
```
I can add a cast if you like.

cfriedt committed:
include: drivers: gpio: use __unused instead of void cast

The __unused attribute is the preferred way of marking
a variable as unused when assertions are disabled.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
Compliance check seems to be a false positive

mbolivar-nordic reviewed (approved):
ACK is based on my previous review and the fact that you're stepping up to collaborator here, which reassures me that you'll stick around to handle anything that breaks as a result of this PR ;)

We will need someone with superpowers to override the compliance false positive
